doc/sphinx/users-guide/vcl-grace.rst

```diff
@@ -19,52 +23,194 @@ If you are serving thousands of hits per second the queue of waiting
+  fresh object is available.
+
+Note that the backend fetch happens asynchronously, and the moment the
+new object is in it will replace the one we've already got.
+
+If you do not define your own ``sub vcl_hit``, then the default one is
+used. It looks like this::
 
   sub vcl_hit {
-     if (obj.ttl >= 0s) {
-         // A pure unadulterated hit, deliver it
-         return (deliver);
-     }
-     if (obj.ttl + obj.grace > 0s) {
-         // Object is in grace, deliver it
-         // Automatically triggers a background fetch
-         return (deliver);
-     }
-     // fetch & deliver once we get the result
-     return (fetch);
+       if (obj.ttl >= 0s) {
+            // A pure unadulterated hit, deliver it
+            return (deliver);
+       }
+       if (obj.ttl + obj.grace > 0s) {
+            // Object is in grace, deliver it
+            // Automatically triggers a background fetch
+            return (deliver);
+       }
+       // fetch & deliver once we get the result
+       return (miss);
   }
 
-The grace logic is pretty obvious here. If you have enabled
-:ref:`users-guide-advanced_backend_servers-health` you can check if
-the backend is sick and only serve graced object then. Replace the
-second if-clause with something like this::
+If you follow the code, you see that Varnish delivers graced objects
+while fetching fresh copies, but if grace has expired the clients have to
+wait until a new copy is available.
+
+Misbehaving servers
+~~~~~~~~~~~~~~~~~~~
+
+A key feature of Varnish is its ability to shield you from misbehaving
+web- and application servers.
+
+If you have enabled :ref:`users-guide-advanced_backend_servers-health`
+you can check if the backend is sick and modify the behavior when it
+comes to grace. There are essentially two ways of doing this. You can
+explicitly deliver kept object (that is not within grace) when you see
+that the backend is sick, or you can explicitly `not` serve an expired
+object when you know that the backend is healthy. The two methods have
+slightly different characteristics, as we shall see.
 
-   if (!std.healthy(req.backend_hint) && (obj.ttl + obj.grace > 0s)) {
-         return (deliver);
-   } else {
-         return (fetch);
+In both cases we assume that you avoid inserting objects into the cache
+when you get certain errors from the backend, for example by using the
+following::
+
+  sub vcl_backend_response {
+       if (beresp.status == 503 && bereq.is_bgfetch) {
+            return (abandon);
+       }
+  }
+
+Method 1: When the backend is healthy, use a lower grace value
+==============================================================
+
+Imagine that you have set an object's grace to a high value that you
+wish to use when the backend is sick, for example::
+
+  sub vcl_backend_response {
+       set beresp.grace = 24h;
+       // no keep
+  }
+
+Then you can use the following code as your ``sub vcl_hit``::
+
+   if (std.healthy(req.backend_hint)) {
```

Comment:
general note: we should have VTCs for all the snippets in the docs to make sure that we don't break them.